Initial Turborepo setup with TypeScript, Tailwind, ESlint

sharmayatendra opened the pull request:
# build: Create initial turbo configuration

This commit initiates the intial configuration for the turborepo project which is created using `npx create-turbo` command.

 [Tutorial-link](https://www.youtube.com/watch?v=YQLw5kJ1yrQ&t=163s)

## files

#### eslint-config-custom

- eslint custom configuration packages next, prettier, react, turbo

#### tsconfig
- add compilerOptions in base, nextJS, react-library file

#### ui
- React, TypeScript, eslint dependencies
- Button, Header component

# feat: Initiate Next projects

This commit aims for the creation of NextJS projects using `create-next-app` command.

### next-app
- add node, next, react, TypeScript dependencies
- configure nextJS
- postcss config file
- app, document and others components are added as well
- add tailwind config file
- add compilerOptions in tsconfig file
- add readme file

### next-next-app
- add same files and folders as next-app
- `name` of app is `next-next-app` which is different from `next-app`

# build: Add shared typescript configs

This commit aims for the creation of shared TypeScript configuration across the project.

### next-app
- extends it from nextjs config file
- modify compilerOptions

### tsconfig
- add some more compiler options in `base-config` 
- add compilerOptions in `server-config`

### server
- add  `express` dependencies
- add `nodemon`, `node`, `typescript`, `@types/express` as devDependencies
- add basic express app
- add an interface to it

few changes made to prettier file as well.

# build: Add shared eslint configs

This commit aims for setting up the linting in the project.

### eslint-config-custom
- Add various dependencies(airbnb,airbnb-base etc), plugins(eslint-plugin) and parser(eslint-parser)
- setup linting rules for nextjs project
- setup linting rules for server

### next-app
- project wise setup for linting
- modify the code according to linting rules like use `arrow-functions` component, React must be in scope

### next-next-app
- same configuration applies as of `next-app` 

# build: Add shared tailwind ui

### ui
- add `tailwindcss`, `postcss`, `autoprefixer` devDependencies

### next-app
- add `next-transpile-module` as devDependency
- conifgure `next-app` with `next-transpile-module`
- use `Button` component in the project

turbobot-temp committed:
feat(create-turbo): apply official-starter transform

sharmayatendra committed:
build: Create initial turbo configuration

This commit introduces the initial configuration for the turborepo project. The configuration is set up using the npx create-turbo command, which creates the necessary files and dependencies to get started.

The following changes have been made:

eslint-config-custom: This package includes custom configurations for eslint, tailored for next, prettier, react, and turbo.

tsconfig: Compiler options have been added to the base, nextJS, and react-library files to ensure proper TypeScript compilation.

ui: Dependencies for React, TypeScript, and eslint have been included. Additionally, the Button and Header components have been added.

For further guidance, you can refer to the tutorial link which provides a video tutorial on setting up the turbo configuration.

sharmayatendra committed:
feat: Initiate Next projects

This commit focuses on creating two Next.js projects using the create-next-app command. The projects are named next-app and next-next-app, and they serve as starting points for building Next.js applications.

The following changes have been made:

In next-app:

Dependencies: Node.js, Next.js, React, and TypeScript dependencies have been added to support the project.

Next.js Configuration: The necessary configurations for Next.js have been set up.

PostCSS Config File: A PostCSS configuration file has been included.

Components: The initial set of components, including app, document, and others, have been added.

Tailwind Config File: A configuration file for Tailwind CSS has been included.

tsconfig: Compiler options have been added to the tsconfig file.

readme: Readme file is also added

In next-next-app:

Same files and folders as next-app have been added.

The application's name is specified as next-next-app, distinguishing it from next-app.

sharmayatendra committed:
build: Add shared TypeScript configs

This commit focuses on introducing shared TypeScript configuration settings that can be used across the project. By centralizing these configurations, we ensure consistency and facilitate easier maintenance and updates.

The changes made are as follows:

In next-app:

Extends the TypeScript configuration from the Next.js config file.

Modifies the compilerOptions to align with the project's requirements.
In tsconfig:

Adds additional compiler options to the base-config section.

Adds compilerOptions specific to the server-config section.

In server:

Adds express dependencies, which are required for building server-side functionality.

Adds nodemon, node, typescript, and @types/express as devDependencies, providing the necessary tools for development.

Implements a basic express application.

Defines an interface for the express app.

Additionally, some changes have been made to the prettier configuration file.

These shared TypeScript configurations enhance code consistency and improve the development experience throughout the project. It ensures that all parts of the project follow the same TypeScript settings and enables easier collaboration between team members.

sharmayatendra committed:
build: Add shared eslint configs

This commit focuses on adding shared ESLint configurations to the project, enabling consistent and standardized linting across the codebase. Linting helps improve code quality, identifies potential issues, and enforces coding best practices.

The following changes have been made:

In eslint-config-custom:

Added various dependencies such as airbnb, airbnb-base, and plugins like eslint-plugin.

Configured linting rules specifically tailored for Next.js projects.

Configured linting rules for the server code.
In next-app:

Implemented project-specific linting setup.

Modified the code according to the linting rules.
This includes using arrow-functions components and ensuring that React is properly in scope.

In next-next-app:

Applied the same linting configuration as next-app.

By adding shared ESLint configurations, we ensure consistent code quality, adhere to coding standards, and promote maintainable code across the project. Linting rules help catch common mistakes and improve code readability, facilitating collaboration within the development team.

sharmayatendra committed:
build: Add shared tailwind ui

This commit focuses on incorporating shared Tailwind UI components into the project. Tailwind CSS is a utility-first CSS framework that helps streamline the development of user interfaces.

The following changes have been made:

In the ui section:

Added devDependencies for tailwindcss, postcss, and autoprefixer. These are necessary for using Tailwind CSS in the project.

In next-app:

Added next-transpile-module as a devDependency. This module enables transpilation of specific modules within Next.js projects.

Configured next-app to utilize next-transpile-module.

Utilized the Button component from the shared Tailwind UI in the project.

By incorporating shared Tailwind UI components, we enhance the project's user interface development. Tailwind CSS simplifies styling and offers a wide range of pre-designed components, such as buttons, that can be easily integrated into the project.

Sowed commented on apps/next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,122 @@
+
+const inter = Inter({ subsets: ['latin'] })
+
+const Home = () => (
```
Use function definition here.

yatensharma replied:
fixed this

Sowed commented on apps/next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,122 @@
+import React from "react"
+import Image from 'next/image'
+import { Button } from 'ui/Button'
```
Separate app/package imports after absolute imports and pad them with a new line.

yatensharma replied:
done

Sowed commented on apps/next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,122 @@
+import React from "react"
```
React import is not required

yatensharma replied:
removed unecessary imports

Sowed commented on apps/next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,122 @@
+        </div>
+      </div>
+
+      <div className="relative flex place-items-center before:absolute before:h-[300px] before:w-[480px] before:-translate-x-1/2 before:rounded-full before:bg-gradient-radial before:from-white before:to-transparent before:blur-2xl before:content-[''] after:absolute after:-z-20 after:h-[180px] after:w-[240px] after:translate-x-1/3 after:bg-gradient-conic after:from-sky-200 after:via-blue-200 after ...
```
Split this into a new line. Use the editor ruler for visual assistance (`"editor.rulers": [80, 120],`)

Sowed replied:
Do this for all lines beyond the first editor ruler point.

yatensharma replied:
configured

Sowed commented on apps/next-next-app/next.config.js:
```diff
@@ -0,0 +1,5 @@
+const withTM = require('next-transpile-modules')(['ui'])
```
Which version of nextjs are you using. This package is now baked into the Next config. 
- [Refer to this](https://github.com/martpie/next-transpile-modules/issues/280#issuecomment-1292615030)
- [transpilePackages](https://nextjs.org/docs/app/api-reference/next-config-js/transpilePackages)

yatensharma replied:
updated

Sowed commented on apps/next-next-app/src/pages/_app.tsx:
```diff
@@ -0,0 +1,10 @@
+import '../styles/globals.css'
+import React from "react"
```
Note, always import 3rd party packages first. And this react import is not necessary above React 18. Move the styles import below and pad them with a new line.

Sowed commented on apps/next-next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,120 @@
+
+const inter = Inter({ subsets: ['latin'] })
+
+const Home = () => (
```
Use function delartion

Sowed commented on apps/next-next-app/src/pages/_document.tsx:
```diff
@@ -0,0 +1,14 @@
+import React from "react"
+import { Html, Head, Main, NextScript } from 'next/document'
+
+const Document = () => (
```
Use function delartion

Sowed commented on apps/next-next-app/src/pages/_app.tsx:
```diff
@@ -0,0 +1,10 @@
+import React from "react"
+import type { AppProps } from 'next/app'
+
+const App = ( {Component, pageProps }: AppProps) => (
```
Use function delartion

Sowed commented on apps/next-next-app/src/pages/index.tsx:
```diff
@@ -0,0 +1,120 @@
+        </div>
+      </div>
+
+      <div className="relative flex place-items-center before:absolute before:h-[300px] before:w-[480px] before:-translate-x-1/2 before:rounded-full before:bg-gradient-radial before:from-white before:to-transparent before:blur-2xl before:content-[''] after:absolute after:-z-20 after:h-[180px] after:w-[240px] after:translate-x-1/3 after:bg-gradient-conic after:from-sky-200 after:via-blue-200 after ...
```
Refer to the previous comment on long lines. Update all lines here to fix that.

Sowed commented on apps/server/package.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "server",
```
Used scoped packages here, and in the rest of the packages and apps.

yatensharma replied:
scoped them

Sowed commented on packages/eslint-config-custom/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "eslint-config-custom",
```
Scope the package

Sowed reviewed (commented):
Hey, I left you some comments but left out most as they will eventually be caught by the linter in the next task.

sharmayatendra committed:
refactor: function declaration, package imports, scoped packages etc.

This commit focuses on refactoring various files by replacing arrow functions with function declarations and introducing scoped packages. Additionally, improvements have been made to package imports, transpiling modules, and code consistency using editor ruler settings.

Server:
Scoped the package to enhance modularity and prevent naming conflicts.

Next-app:
Replaced arrow functions with function declarations for improved readability and maintainability.

Scoped packages to ensure encapsulation and prevent naming clashes.

Introduced transpile modules to optimize module bundling.

Utilized editor ruler settings to maintain consistent code formatting and adherence to style guidelines.

UI (Design system):
Renamed "ui" to "design-system" to accurately reflect its purpose and intent.

Transitioned from arrow functions to function declarations for code consistency and standardization.

Scoped packages to promote encapsulation and minimize potential conflicts.

By making these refactorings, the codebase becomes more organized, easier to read, and aligned with established best practices. These changes improve code maintainability, enhance modularity, and contribute to a more robust and scalable project structure.
